clippy_lints/src/methods/mod.rs

```diff
@@ -2293,6 +2294,83 @@ declare_clippy_lint! {
     more clearly with `if .. else ..`"
 }
 
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for usage of:
+    ///
+    /// - `[foo].iter()`
+    /// - `[foo].iter_mut()`
+    /// - `[foo].into_iter()`
+    /// - `Some(foo).iter()`
+    /// - `Some(foo).iter_mut()`
+    /// - `Some(foo).into_iter()`
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It is simpler to use the once function from the standard library:
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// let a = [123].iter();
+    /// let b = Some(123).into_iter();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::iter;
+    /// let a = iter::once(&123);
+    /// let b = iter::once(123);
+    /// ```
+    ///
+    /// ### Known problems
+    ///
+    /// The type of the resulting iterator might become incompatible with its usage
+    #[clippy::version = "1.64.0"]
+    pub ITER_ONCE,
+    pedantic,
+    "Iterator for array of length 1"
+}
+
+declare_clippy_lint! {
+    /// ### What it does
+    ///
+    /// Checks for usage of:
+    ///
+    /// - `[].iter()`
+    /// - `[].iter_mut()`
+    /// - `[].into_iter()`
+    /// - `None.iter()`
+    /// - `None.iter_mut()`
+    /// - `None.into_iter()`
+    ///
+    /// ### Why is this bad?
+    ///
+    /// It is simpler to use the empty function from the standard library:
+    ///
+    /// ### Example
+    ///
+    /// ```rust
+    /// use std::{slice, option};
+    /// let a: slice::Iter<i32> = [].iter();
+    /// let f: option::IntoIter<i32> = None.into_iter();
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::iter;
+    /// let a: iter::Empty<i32> = iter::empty();
+    /// let b: iter::Empty<i32> = iter::empty();
+    /// ```
+    ///
+    /// ### Known problems
+    ///
+    /// The type of the resulting iterator might become incompatible with its usage
```

Comment:
I'm a little concerned about this. From how many allows are necessary in Clippy already, I think this is a bigger problem, that has to be addressed before we can merge this.

Reply:
Does clippy have a way to know whether a suggestion type-checks?
Maybe we could limit it to expressions within a `chain`, which would likely lead to less false positives but not 0. It would also create false negatives in cases like:

```rust
let iter = if some_condition {
    [a].into_iter()
} else {
    [b].into_iter()
}
  
iter.chain(other_iter)
```

Reply:
Do you have examples where applying the suggestion does not type check? Looking at all the `allow`s you added to Clippy, I don't really see why you couldn't use `iter::empty` there.

Reply:
Hm, if I'd had to guess, I would say that this fails type checking when using `iter::empty` because it can't infer the type anymore? I don't think this is a problem for `iter_once` though.

Reply:
Say you have something like this:

```rust
let iter = match smth {
    "a" => [1,2].iter(),
    "b" => [3].iter(),
    _ => [].iter(),
};
```

The first branch forces `iter` to be of type `slice::Iter<i32>` but suggestions would be to change it to:

```rust
let iter = match smth {
    "a" => [1,2].iter(),
    "b" => iter::once(&3),
    _ => iter::empty(),
};
```

so now `iter` must be at the same time `Once<i32>`, `slice::Iter<i32>` and `Empty<i32>`

Reply:
Hm, limiting this to iter chains also doesn't solve this, e.g.:

```rust
let iter = match smth {
    "a" => [[1],[2]].iter().flatten(),
    "b" => [[3]].iter().flatten(),
    _ => [].iter().flatten(),
};
    
let iter = match smth {
    "a" => [[1],[2]].iter().flatten(),
    //"b" => [[3]].iter().flatten(),
    "b" => std::iter::once([3]).flatten(),
    _ => std::iter::empty().flatten(),
};
```

[Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=e6b3994fbecbb6b4ab4d9a369c9328db)

I don't really have a solution to this... I like the idea of the lint, but think that this is too big of a FP to have it as a `pedantic` lint. I'll have to check how often this FP appears in the wild.

Reply:
Running it on the top 50 crates, those lints were only emitted on `serde_derive` and all 3 cases were fixable. Maybe I making a mountain out of a molehill here.

Reply:
I asked on Zulip about this https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/.60iter_empty.60.2F.60iter_once.60.20new.20lints.20clippy.239187/near/290350623

Reply:
> Running it on the top 50 crates, those lints were only emitted on serde_derive and all 3 cases were fixable. Maybe I making a mountain out of a molehill here.

@flip1995 have you ensured that the lint was enabled when you ran it on those crates, or could it be that they haven't enabled `pedantic` lints?

Reply:
I used `@mikerite`s tool to check those lints. I'm pretty sure they were enabled. I mean one of those triggered on serde_derive.

Reply:
I ran the tool on the top 200 crates to have a better idea.


The results are:
| Crate                        | Count | False Positive (FP)| Partial FP |
|:-----------------------------|:-----:|:-----:             | :--------: |
| chrono-0.4.19                |     1 |               0     |0  |
| petgraph-0.6.2               |     2 |                   1|  0|
| rayon-1.5.3                  |     1 |                   1|  0|
| serde_derive-1.0.140         |     3 |                   0| 0|
| synstructure-0.12.6          |     1 |                   0| 0 |
| tokio-1.20.0                 |     2 |                   0| 2          |
| unicode-normalization-0.1.21 |     6 |                   6| 1          |
| clippy | 7 | 2 | 0 |

*Partial FP* means the lint is right to trigger but the suggestion doesn't typecheck.

#### [petgraph](https://github.com/petgraph/petgraph/blob/master/src/graphmap.rs#L355=) and  [rayon](https://github.com/rayon-rs/rayon/blob/master/src/vec.rs#L279=) 

False positive, a `slice::Iter(Mut)` is needed

#### `serde_derive` ([1](https://github.com/serde-rs/serde/blob/master/serde_derive/src/bound.rs#L344=) [2](https://github.com/serde-rs/serde/blob/master/serde_derive/src/de.rs#L3001=) [3](https://github.com/serde-rs/serde/blob/master/serde_derive/src/de.rs#L3067=)) and [chrono](https://github.com/chronotope/chrono/blob/main/src/datetime/mod.rs#L622=)

Each time the perfect true positive for this lint

#### [tokio](https://github.com/tokio-rs/tokio/blob/master/tokio/src/net/addr.rs#L57=)

Two "false positive". In the sense that the suggestion doesn't TypeCheck but by changing some type annotations it does. This could have required a breaking change if it hadn't been on some private `impls` though.

#### [unicode-normalization](https://github.com/unicode-rs/unicode-normalization/blob/master/src/lib.rs#L173=) 
6 times the same false positive for type checking, because a trait is implemented specifically for `Option::IntoIter`, but it might be interesting for them to implement the trait also for `iter::Once`. I don't know how the crate is used and whether it would be useful though.
 
#### Clippy itself (in this PR)

The lint is fixed 5 times and ignored twice. There was one "partial false positive" which required manually writing the type of `Empty<ty::subst::GenericArg<'_>>`, which was a bit of a pain.

Reply:
3b45a28090ea95a6ad0a589a581a1297070521cc based on [this](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/.60iter_empty.60.2F.60iter_once.60.20new.20lints.20clippy.239187/near/290379301) fixed the FP in petgraph and clippy itself. The Lint still triggers in Tokio and unicode-normalization though.